compiler/circle-execution-plan/src/ExecutionOrderEstimator.cpp

```diff
@@ -0,0 +1,235 @@
+    j--;
+
+  if (j == -1)
+    return false;
+
+  int k = n - 1;
+  while (vector[j] >= vector[k])
+    k--;
+
+  std::swap(vector[j], vector[k]);
+  int l = j + 1, r = n - 1;
+
+  while (l < r)
+    std::swap(vector[l++], vector[r--]);
+
+  return true;
+}
+
+} // namespace
+
+void ExecutionOrderEstimator::get_node_order_inform_vector(
+  std::vector<NodeOrderInform> &node_order_inform_vector)
+{
+  std::set<loco::Node *> visited_nodes;
+  std::stack<Frame> frames;
+
+  auto visited = [&visited_nodes](loco::Node *node) {
+    return visited_nodes.find(node) != visited_nodes.end();
+  };
+
+  for (auto node : _root_nodes)
+  {
+    assert((node != nullptr) && "root is invalid");
+    frames.push(Frame{node});
+  }
+
+  while (!frames.empty())
+  {
+    auto &top_frame = frames.top();
+
+    if (top_frame.pos() == -1)
+    {
+      if (visited(top_frame.ptr()))
+      {
+        frames.pop();
+        continue;
+      }
+      visited_nodes.insert(top_frame.ptr());
+
+      auto exec_nodes_count = 0;
+      for (uint32_t i = 0; i < top_frame.ptr()->arity(); ++i)
+      {
+        if (auto next = top_frame.ptr()->arg(i))
+          if (isExecutableNode(loco::must_cast<luci::CircleNode *>(next)))
+            exec_nodes_count++;
+      }
+      if (exec_nodes_count > 1)
+      {
+        NodeOrderInform node_order_inform;
+        node_order_inform._node = top_frame.ptr();
+        node_order_inform_vector.emplace_back(node_order_inform);
+      }
+    }
+
+    top_frame.advance();
+
+    assert(top_frame.pos() >= 0);
+
+    auto it = std::find_if(node_order_inform_vector.begin(), node_order_inform_vector.end(),
+                           [top_frame](const auto &node_order_inform) {
+                             return node_order_inform._node == top_frame.ptr();
+                           });
+
+    if (top_frame.pos() < static_cast<int64_t>(top_frame.ptr()->arity()))
+    {
+      if (auto next = top_frame.ptr()->arg(top_frame.pos()))
+      {
+        if (it != node_order_inform_vector.end())
+        {
+          if (isExecutableNode(loco::must_cast<luci::CircleNode *>(next)))
+          {
+            it->executable_arg_idx.emplace_back(top_frame.pos());
+          }
+          else
+          {
+            it->non_executable_arg_idx.emplace_back(top_frame.pos());
+          }
+        }
+        frames.push(Frame{next});
+      }
+    }
+    else
+    {
+      // Let's visit the current argument (all the arguments are already visited)
+      if (it != node_order_inform_vector.end())
+      {
+        do
+        {
+          auto combination = it->executable_arg_idx;
+          for (unsigned int &s : it->non_executable_arg_idx)
+            combination.emplace_back(s);
+          it->all_combinations.emplace_back(combination);
+        } while (nextPermutation(it->executable_arg_idx));
+      }
+      frames.pop();
+    }
+  }
+}
+
+std::vector<loco::Node *> ExecutionOrderEstimator::get_current_order(
```

Comment:
Q) I've just read the rest of code and noticed that it not producing sequences like this:
![order](https://user-images.githubusercontent.com/3631351/146770483-e4904935-bb20-4ff9-abb0-8d2fdd30d84b.png)

Did I understand this algorithm right?

Reply:
No, the sequence for the graph you mentioned above could be: 
- 1, 2, 4, 3, 5, 6
- 1, 3, 5, 2, 4, 6
We only estimate the order of branches

Reply:
Then, I want to take my words back =)
We need to name this search algorithm different. "Exhaustive search" implies that it checks all possible sequences